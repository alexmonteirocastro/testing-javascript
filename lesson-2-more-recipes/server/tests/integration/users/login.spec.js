import supertest from 'supertest';
import bcrypt from 'bcrypt';
import app from '../../../index';
import { User } from '../../../database/models'

describe('The user login', () => {

    beforeEach( async () => {
        // Clear the database
        await User.destroy({ where: {} })
    });
    
    test('the user can log in and get a jwt ', async () => {
        
        /// arrange

        //setup fake user data
        const fakeUser = {
            name: 'bahdcoder',
            email: 'bahdcoder@gmail.com',
            password: 'password'
        }

        // create new user
        await User.create({
            name: fakeUser.name,
            email: fakeUser.email,
            password: bcrypt.hashSync(fakeUser.password, 1)
        }); //come back later
        
        /// action

        // make POST request to login
        const response = await supertest(app).post('/api/v1/users/signin').send({
            email: fakeUser.email,
            password: fakeUser.password
        })

        //console.log(response.body)


        /// assertion
        expect(response.status).toBe(200);

        //assert response from server contains jwt and user data
        expect(response.body.data.access_token).toBeTruthy();
        expect(response.body.data.user.email).toBe(fakeUser.email);


    });

    test('if user tries to login without password, return error password required', async () => {
        
        /// Arrange

        //setup fake user data
        const fakeUser = {
            name: 'bahdcoder',
            email: 'bahdcoder@gmail.com',
            password: ''
        }

        // create new user
        await User.create({
            name: fakeUser.name,
            email: fakeUser.email,
            password: fakeUser.password
        }); //come back later
        
        /// action

        // make POST request to login
        const response = await supertest(app).post('/api/v1/users/signin').send({
            email: fakeUser.email,
            password: fakeUser.password
        })

        // console.log(response.body)

        /// Assertion
        expect(response.body.status).toBe('fail');
        expect(response.body.data).toEqual({errors: ['The password is required.']});

    });

    test('if user tries to login without email, return error email required', async () => {
        /// Arrange

        //setup fake user data
        const fakeUser = {
            name: 'bahdcoder',
            email: '',
            password: 'password'
        }

        // create new user
        await User.create({
            name: fakeUser.name,
            email: fakeUser.email,
            password: fakeUser.password
        }); //come back later
        
        /// action

        // make POST request to login
        const response = await supertest(app).post('/api/v1/users/signin').send({
            email: fakeUser.email,
            password: fakeUser.password
        })

        //console.log(response.body)

        /// Assertion
        expect(response.body.status).toBe('fail');
        expect(response.body.data).toEqual({errors: ['The email is required.']});
    });

    test('if user tries to login with invalid email, return error email must be valid', async () => {
        
        /// Arrange

        //setup fake user data
        const fakeUser = {
            name: 'bahdcoder',
            email: 'bahdcoder@gmailcom',
            password: 'password'
        }

        // create new user
        await User.create({
            name: fakeUser.name,
            email: fakeUser.email,
            password: fakeUser.password
        }); //come back later
        
        /// action

        // make POST request to login
        const response = await supertest(app).post('/api/v1/users/signin').send({
            email: fakeUser.email,
            password: fakeUser.password
        })

        // console.log(response)

        /// Assertion
        expect(response.body.status).toBe('fail');
        expect(response.body.data).toEqual({errors: ['The email must be a valid email address.']});
    });

});
