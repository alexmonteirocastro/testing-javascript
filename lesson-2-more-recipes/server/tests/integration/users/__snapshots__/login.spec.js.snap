// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The user login if user tries to login with invalid email, return error email must be valid 1`] = `
Object {
  "header": Object {
    "access-control-allow-origin": "*",
    "connection": "close",
    "content-length": "80",
    "content-type": "application/json; charset=utf-8",
    "date": "Thu, 05 Jul 2018 13:38:17 GMT",
    "etag": "W/\\"50-ApG9uD1xEep3TQHaYI4BQBuXHgU\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": Object {
      "email": "bahdcoder@gmailcom",
      "password": "password",
    },
    "headers": Object {
      "content-type": "application/json",
      "user-agent": "node-superagent/3.8.2",
    },
    "method": "POST",
    "url": "http://127.0.0.1:35509/api/v1/users/signin",
  },
  "status": 422,
  "text": "{\\"status\\":\\"fail\\",\\"data\\":{\\"errors\\":[\\"The email must be a valid email address.\\"]}}",
}
`;
